feat(forge): revert diagnostic inspector

0xrusowsky opened the pull request:
## closes
- https://github.com/foundry-rs/foundry/issues/4141

## design

Created a new `RevertDiagnostic` inspector that provides detailed information of otherwise undefined evm reverts. For the time being, the inspector suports:
```rs
#[derive(Debug, Clone, Copy)]
pub enum DetailedRevertReason {
    CallToNonContract(Address),
    DelegateCallToNonContract(Address),
}
```
Additionally, enhanced the traces decoder to determine when a function call to an unkown selector takes places (only for local contracts with artifacts).


## output
these would be the new error messages that users would see:
```
Traces:
  [6350] NonContractCallRevertTest::test_non_contract_call_failure()
    ├─ [0] console::log("test non contract call failure") [staticcall]
    │   └─ ← [Stop]
    ├─ [0] 0xdEADBEeF00000000000000000000000000000000::number()
    │   └─ ← [Stop]
    └─ ← [Revert] call to non-contract address 0xdEADBEeF00000000000000000000000000000000
    // rest of the logs

Traces:
  [255303] NonContractDelegateCallRevertTest::test_unlinked_library_call_failure()
    ├─ [0] console::log("Test: Simulating call to unlinked library") [staticcall]
    │   └─ ← [Stop]
    ├─ [214746] → new LibraryCaller@0x5615dEB798BB3E4dFa0139dFa1b3D433Cc23b72f
    │   └─ ← [Return] 960 bytes of code
    ├─ [3896] LibraryCaller::foobar(10)
    │   ├─ [0] 0xdEADBEeF00000000000000000000000000000000::foo(10) [delegatecall]
    │   │   └─ ← [Stop]
    │   └─ ← [Revert] delegatecall to non-contract address 0xdEADBEeF00000000000000000000000000000000 (usually an unliked library)
    // rest of the logs

Traces:
  [8620] NonContractCallRevertTest::test_non_supported_selector_call_failure()
    ├─ [0] console::log("test non supported fn selector call failure") [staticcall]
    │   └─ ← [Stop]
    ├─ [145] Counter::random()
    │   └─ ← [Revert] unrecognized function selector 0x5ec01e4d for contract 0x5615dEB798BB3E4dFa0139dFa1b3D433Cc23b72f, which has no fallback function.
    // rest of the logs
```

## benchmarks `forge test -vvvv`

| Project | Before [[v1.2.1]](https://github.com/foundry-rs/foundry/releases/tag/v1.2.1-rc) | After | Overhead |
|--------|----------|------|-----------|
|[morpho-blue](https://github.com/morpho-org/morpho-blue)| 1.93s | 2.14s | 10.8% |
|[spark-psm](https://github.com/sparkdotfi/spark-psm)| 39.56s | 44.88s | 13.4% |
|[uniswap/v4-core](https://github.com/Uniswap/v4-core)| 6.93s | 7.72s | 11.4% |
|[vectorized/solady](https://github.com/Vectorized/solady)| 2.44s | 2.67s | 9.4% |

**notes**:
- benchmarks only measure running of tests and are pre-build
- benchmarks were run on an m2pro with 16gb of RAM
- each test run 5 times and the reported time is the avg

**conclusion:** the new inspector makes tests (on average) ~11% slower... the bright side is that it would only impact those test runs where traces are active (which i would assume to only be when actively debugging).

do we feel like this is acceptable?

grandizzy commented on crates/evm/evm/src/inspectors/revert_diagnostic.rs:
```diff
@@ -0,0 +1,100 @@
+}
+
+impl RevertDiagnostic {
+    fn is_delegatecall(&self, scheme: CallScheme) -> bool {
```
these can be all helper functions, not RevertDiagnostic specific

0xrusowsky replied:
would u keep them in the inspector file? or should i place them somewhere else (like utils or similar)?

grandizzy commented on crates/evm/evm/src/inspectors/revert_diagnostic.rs:
```diff
@@ -0,0 +1,100 @@
+            return None;
+        }
+
+        let reason = match self.non_contract_call {
```
can use `self.non_contract_call.map`

grandizzy commented on crates/evm/evm/src/inspectors/stack.rs:
```diff
@@ -61,6 +61,8 @@ pub struct InspectorStackBuilder {
     /// The CREATE2 deployer address.
     pub create2_deployer: Address,
+    /// Whether to provide diagnostics for EVM reverts.
+    pub revert_diag: Option<bool>,
```
it makes sense to always provide such, wdyt?

0xrusowsky replied:
just wanted to keep the existing pattern, but i'd do the same 👍

grandizzy commented on crates/forge/tests/cli/test_cmd.rs:
```diff
@@ -3600,3 +3600,152 @@ Ran 1 test suite [ELAPSED]: 2 tests passed, 0 failed, 0 skipped (2 total tests)
+
+Failing tests:
+Encountered 1 failing test in test/NonContractCallRevertTest.t.sol:NonContractCallRevertTest
+[FAIL: EvmError: call to non-contract address `0xdEADBEeF00000000000000000000000000000000`] test_non_contract_call_failure() ([GAS])
```
nice!

grandizzy commented on crates/evm/evm/src/inspectors/revert_diagnostic.rs:
```diff
@@ -0,0 +1,100 @@
+        outcome: CallOutcome,
+    ) -> CallOutcome {
+        if outcome.result.result == InstructionResult::Revert {
+            self.reverted = true
```
maybe here we could directly set the reason? we can also do more checks if target address is actually a contract, (like checking if selector available, etc. some other similar hh checks https://github.com/NomicFoundation/hardhat/blob/67f1e95e1f3904f7b2e8a5560115c1551e899f64/packages/hardhat-core/src/internal/hardhat-network/stack-traces/solidity-errors.ts#L218-L341)

klkvr replied:
I think we need some smarter strategy here to ensure that call we've caught in `call` hook is indeed the one that caused revert (e.g compare depth, ensure there were no calls after it)

grandizzy reviewed (commented):
IMO on the right track, @klkvr pls share your thoughts re the approach.
left some minor comments / nits.
To note that such default option could add perf penalty for invariant tests running with `fail_on_revert = false` but probably bearable.

klkvr commented on crates/forge/tests/cli/test_cmd.rs:
```diff
@@ -3600,3 +3600,152 @@ Ran 1 test suite [ELAPSED]: 2 tests passed, 0 failed, 0 skipped (2 total tests)
+import "forge-std/Test.sol";
+import {Counter} from "../src/Counter.sol";
+
+interface ICounter {
+    function number() external returns (uint256);
+}
```
It seems that in cases of calling functions returning data Solidity's approach is to revert if data is empty.

However, if method does not return anything (e.g like `increment()`) then Solidity would firstly check before the call whether the address has any code, and revert if it's not which I believe is not being caught by our inspector rn

0xrusowsky commented:
addresses the PR feedback by checking the call stack depth + tracking EXTCODESIZE calls to empty addresses (confirmed that the initial impl wouldn't diagnose those).

the only downside is extra performance overhead, but i tried to make the `fn setp` as efficient as possible (improvement suggestions are welcome)

0xrusowsky commented:
i also enhanced the trace decoder to identify function calls for a selector that is not supported by the called contract (if the tracer has access to its abi)

grandizzy commented on crates/evm/evm/src/inspectors/revert_diagnostic.rs:
```diff
@@ -0,0 +1,132 @@
+        if EXTCODESIZE == interpreter.current_opcode() {
+            if let Ok(word) = interpreter.stack().peek(0) {
+                let addr = Address::from_word(word.into());
+                if let Ok(state) = ctx.code(addr) {
```
this warms the account (`load_account`) hence the gas changes in tests, we should probably just record these addresses and depth and check for the code in call_end, if call reverted

grandizzy reviewed (approved):
makes sense, thank you! pending others review before merging
CC @klkvr

zerosnacks reviewed (approved):
lgtm :+1: pending final review from @klkvr 

Note: there will likely be a small merge conflict after I merge https://github.com/foundry-rs/foundry/pull/10183

0xrusowsky commented:
> lgtm 👍 pending final review from @klkvr
> 
> Note: there will likely be a small merge conflict after I merge #10183

great, thanks!
np, i'll take care of that once it shows up 🤝

klkvr commented on crates/evm/evm/src/inspectors/revert_diagnostic.rs:
```diff
@@ -0,0 +1,147 @@
+    pub reverted: bool,
+}
+
+impl RevertDiagnostic {
```
could we please add docs to all methods here

klkvr commented on crates/evm/evm/src/inspectors/revert_diagnostic.rs:
```diff
@@ -0,0 +1,147 @@
+            }
+        }
+
+        if let Some((addr, _)) = self.non_contract_size_check {
+            // call never took place, so schema is unknown --> output most generic msg
+            return Some(DetailedRevertReason::CallToNonContract(addr));
+        }
```
this is reachable even if we did not revert, is it expected?

klkvr commented on crates/evm/evm/src/inspectors/revert_diagnostic.rs:
```diff
@@ -0,0 +1,147 @@
+
+    /// Records whether the call reverted or not. Only if the revert call depth matches the
+    /// inspector cache.
+    fn call_end(
+        &mut self,
+        ctx: &mut EvmContext<DB>,
+        _inputs: &CallInputs,
+        outcome: CallOutcome,
+    ) -> CallOutcome {
+        if let Some((_, _, depth)) = self.non_contract_call {
+            if outcome.result.result == InstructionResult::Revert &&
+                ctx.journaled_state.depth() == depth - 1
+            {
+                self.reverted = true
+            };
+        }
+
+        outcome
+    }
```
I would like us to reduce false positives as much as possible here. I believe such Solidity reverts always have empty revert reason so we could catch this here

I'm also wondering if instead of propagating the `maybe_reason` all the way up to the `RevertDecoder` we could just manually replace the revert reason here with something more meaningful? an example of this is here https://github.com/foundry-rs/foundry/blob/959c713539d30ce787a8d1fe84c44bbd13bc90e4/crates/cheatcodes/src/test/expect.rs#L816-L823

One benefit of such approach would be that we'd be able to override revert reasons on any depth vs right now we can only decode a single such error in the trace

klkvr commented on crates/evm/traces/src/decoder/mod.rs:
```diff
@@ -364,6 +371,45 @@ impl CallTraceDecoder {
                 }
             };
+
+            // Check if unsupported fn selector: calldata dooes NOT point to one of its selectors +
+            // non-fallback contract + no receive
+            if let Some(contract_selectors) = self.non_fallback_contracts.get(&trace.address) {
+                if !contract_selectors.contains(&selector) &&
+                    (!cdata.is_empty() || !self.receive_contracts.contains(&trace.address))
+                {
+                    let return_data = if !trace.success {
+                        let revert_msg =
+                            self.revert_decoder.decode(&trace.output, Some(trace.status), None);
```
this will only work for local contracts right? I'm wondering if there's some common bytecode pattern that is being reached when contract is called with such wrong calldata which we could catch and make this logic work for any contract for which we don't know abi

0xrusowsky replied:
yes, exactly this would only provide insights if we know the abi beforehand

i guess a generic approach would be to perform some sort bytecode analysis and identify when the PC doesn't do any jumps on the function dispatcher and ends up reverting (idk if this vyper uses the same approach as solidity though)

klkvr reviewed (changes requested):
very nice! overall the approach lgtm, this feature is kind of tricky because we're basically forced to do bytecode/execution pattern matching and I'd like us to reduce false positives here and make sure we document this clearly because this logic will likely get more complex over time

0xrusowsky commented:
@klkvr i've addressed all the feedback that u left:
- improved docs
- made things more restrictive by enforcing "empty revert data" as a condition
- injected the revert reason into the inspector --> to do this, i had to move the `fn call_end()` logic into `fn step()`

lmk if i should fix something else

PS: as discussed, we'd leave the trace decoder improvement (where we would analyze that there are no matches in the fn dispatcher) for another PR

klkvr commented on crates/evm/evm/src/inspectors/revert_diagnostic.rs:
```diff
@@ -0,0 +1,213 @@
+    fn call(&mut self, ctx: &mut CTX, inputs: &mut CallInputs) -> Option<CallOutcome> {
+        let target = self.code_target_address(inputs);
+
+        if IGNORE.contains(&target) || self.is_precompile(ctx.cfg().spec().into(), target) {
```
we can do `ctx.journal_ref().precompile_addresses().contains(&target)` here instead which should be cheaper

klkvr commented on crates/evm/evm/src/inspectors/revert_diagnostic.rs:
```diff
@@ -0,0 +1,213 @@
+            return None;
+        }
+
+        if let Ok(state) = ctx.journal().code(target) {
+            if state.is_empty() && !inputs.input.is_empty() {
+                self.non_contract_call = Some((target, inputs.scheme, ctx.journal().depth()));
+            }
+        }
```
note: pattern `CALL -> REVERT` actually means that Solidity was not able to decode outputs which might happen even when contract has code but reported with invalid abi encoding like in e.g common case of ERC20 transfer not returning bool for some tokens. Ideally we should also be able to catch this case but we can track this separately

0xrusowsky replied:
i'll open a new issue for that one and the unsupported selector.

just to be certain that i don't miss anything:
1. **missmatch between bytecode and assumed abi:** we can catch it cause there is a successful call, followed by a revert at the same depth where the call took place --> to avoid false positives, we should know check the returndata validation logic, but that requires careful inspection of the opcodes following the call. any hints on what u'd do exactly?
2. **call to non supported fn selector:** we can catch it (at least assuming that the target is a solidity/vyper contract) if, when a call takes places, there is a revert without any previous positive `JUMPI`, as that would mean that the PC walks through the whole fn dispatcher without finding a match --> using the traces decoder is limited to local contracts, but will never throw false positives. with this logic we could do everything with the inspector alone, which i like. however, do you think the approach is sound?

klkvr replied:
> to avoid false positives, we should know check the returndata validation logic, but that requires careful inspection of the opcodes following the call. any hints on what u'd do exactly?

I think it's likely that Solidity generates some common chunk of code that's always reached in such cases. you could try writing a contract that has many such calls that are failing to decode the calldata in different methods. I'd expect all of those calls to end up in a similar `REVERT` opcode which we can then somehow identify and use to catch such cases

> we can catch it (at least assuming that the target is a solidity/vyper contract) if, when a call takes places, there is a revert without any previous positive JUMPI, as that would mean that the PC walks through the whole fn dispatcher without finding a match 

not sure if the JUMPI approach would work because iirc there are some checks around call value also, we could similarly try to see how the fallbackless contract revert looks like, I'd expect Solidity to generate similar bytecode for such

klkvr commented on crates/evm/evm/src/inspectors/revert_diagnostic.rs:
```diff
@@ -0,0 +1,213 @@
+        // REVERT (offset, size)
+        if REVERT == interp.bytecode.opcode() {
+            if let Ok(size) = interp.stack.peek(1) {
+                if size == U256::ZERO {
```
```suggestion
                if size.is_zero() {
```

klkvr reviewed (approved):
lgtm, only have small nits

grandizzy reviewed (approved):
thank you, good to send it from my pov! We should make some comparison with the project we use to have as a baseline (see https://github.com/foundry-rs/foundry/pull/10183#issue-2950078730) to be aware of penalty it introduce when forge test with traces.

@klkvr @zerosnacks @DaniPopes last pass through before merge please, thank you!

DaniPopes commented on crates/evm/evm/src/inspectors/revert_diagnostic.rs:
```diff
@@ -0,0 +1,207 @@
+    /// When an `EXTCODESIZE` opcode occurs:
+    ///  - Optimistically caches the target address and current depth in `non_contract_size_check`,
+    ///    pending later validation.
+    fn step(&mut self, interp: &mut Interpreter, ctx: &mut CTX) {
```
Please remember that `step` and `step_end` are **INCREDIBLY hot functions**, so any minimal change has a huge effect on execution performance

Please look at cheatcodes inspector on how to outline the if conditions inside of step and step_end; in this case you should do something like this:
```
match opcode {
    op::REVERT => self.handle_revert(...),
    op::EXTCODESIZE => self.handle_extcodesize(...),
    _ => {}
}

#[cold]
fn handle_* ...
```

DaniPopes commented on crates/evm/evm/src/inspectors/revert_diagnostic.rs:
```diff
@@ -0,0 +1,207 @@
+                if IGNORE.contains(&addr) ||
+                    ctx.journal_ref().precompile_addresses().contains(&addr)
+                {
+                    return;
```
these `return` are incorrect if more checks are added in this function, outside of the `if`; however if you extract to separate functions as mentioned in the other comment these are fine

DaniPopes commented on crates/evm/evm/src/inspectors/revert_diagnostic.rs:
```diff
@@ -0,0 +1,207 @@
+
+    /// Tracks `EXTCODESIZE` output. If the bytecode size is 0, clears the cache.
+    fn step_end(&mut self, interp: &mut Interpreter, _ctx: &mut CTX) {
+        if self.is_extcodesize_step {
```
same here as in `step`
